pkg/catalog/catalog.go

```diff
@@ -991,23 +990,75 @@ func (c *Catalog) DiffUncommitted(ctx context.Context, repository string, branch
 	}
 	it := NewEntryDiffIterator(iter)
 	defer it.Close()
-	return listDiffHelper(it, limit, after)
+	return listDiffHelper(it, prefix, delimiter, limit, after)
+}
+
+func GetStartPos(prefix, after, delimiter string) string {
+	// find the correct place to start iterating from based on delimiter, prefix, after
+	switch {
+	case prefix == "" && after == "":
+		return ""
+	case delimiter != "" && after != "":
+		return string(graveler.UpperBoundForPrefix([]byte(after)))
+	case after != "":
+		return after
+	default:
+		return prefix
+	}
 }
 
-func listDiffHelper(it EntryDiffIterator, limit int, after string) (Differences, bool, error) {
+const commonPrefixSplitParts = 2
+
+func listDiffHelper(it EntryDiffIterator, prefix, delimiter string, limit int, after string) (Differences, bool, error) {
 	if limit < 0 || limit > DiffLimitMax {
 		limit = DiffLimitMax
 	}
-	afterPath := Path(after)
-	if afterPath != "" {
-		it.SeekGE(afterPath)
-	}
+	seekStart := GetStartPos(prefix, after, delimiter)
+	it.SeekGE(Path(seekStart))
+
 	diffs := make(Differences, 0)
 	for it.Next() {
 		v := it.Value()
-		if v.Path == afterPath {
-			continue
+		path := string(v.Path)
+
+		if path == after {
+			continue // emulate SeekGT using SeekGE
+		}
+		if !strings.HasPrefix(path, prefix) {
+			break // we only want things that start with prefix, apparently there are none left
 		}
+
+		if delimiter != "" {
+			// common prefix logic goes here.
+			// for every path, after trimming "prefix", take the string upto-and-including the delimiter
+			// if the received path == the entire remainder, add that object as is
+			// if it's just a part of the name, add it as a "common prefix" entry -
+			//   and skip to next record following all those starting with this prefix
```

Comment:
Consider reduce the code block size inside this loop by splitting it to extractCommonPrefix, checking if there is a common prefix, and fall to the part where we append and check the limit)